core/trino-main/src/test/java/io/trino/execution/TestSqlTaskManager.java

```diff
@@ -444,17 +462,40 @@ public URI createMemoryInfoLocation(InternalNode node)
     private static class MockSplitRunner
             implements SplitRunner
     {
-        private SettableFuture<Boolean> interrupted = SettableFuture.create();
+        private final SettableFuture<Void> startedFuture = SettableFuture.create();
+        private final SettableFuture<Void> finishedFuture = SettableFuture.create();
+
+        @GuardedBy("this")
+        private Thread runnerThread;
+        @GuardedBy("this")
+        private boolean closed;
+
+        public void waitForStart()
+                throws ExecutionException, InterruptedException, TimeoutException
+        {
+            startedFuture.get(10, SECONDS);
+        }
+
+        public void waitForFinish()
+                throws ExecutionException, InterruptedException, TimeoutException
+        {
+            finishedFuture.get(10, SECONDS);
+        }
 
         @Override
-        public boolean isFinished()
+        public synchronized boolean isFinished()
         {
-            return interrupted.isDone();
+            return closed;
         }
 
         @Override
         public ListenableFuture<Void> processFor(Duration duration)
         {
+            startedFuture.set(null);
+            synchronized (this) {
```

Comment:
You need to have something like
```
if(closed){
 finishedFuture.set(null);
 return;
}
```
in the synchronized section. Otherwise if the split is closed before processing the thread will never get interrupted